### PR TITLE
fix(ssa): fix the compile-time check for equality in acir-gen

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Summary of changes

Check for equality before creating a witness and a constraint, in case the equality is known at compile-time.

## Test additions / changes

This issue happens with the poseidonsponge_x5_254 test, but it is not part of the CI because it would be too long.

# Checklist

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [X] I have reviewed the changes on GitHub, line by line.
- [X] I have ensured all changes are covered in the description.
- [ ] This PR requires documentation updates when merged.

### CODE REVIEW

File: crates/noirc_evaluator/src/ssa/acir_gen/operations/cmp.rs [hunk 1]
```diff
@@ -64,28 +69,30 @@ pub(crate) fn evaluate_neq(
 
         return Expression::from(&constraints::evaluate_zero_equality(x_witness, evaluator));
     }
+
+    // Arriving here means that `lhs` and `rhs` are not Arrays
     let l_c = l_c.expect("ICE: unexpected array pointer");
     let r_c = r_c.expect("ICE: unexpected array pointer");
```
Comment (kevaundray) [inline crates/noirc_evaluator/src/ssa/acir_gen/operations/cmp.rs:75]: I'm guessing here, we don't need to check if `l_c == r_c` ?

**Review by kevaundray — APPROVED** [review]
This looks good to me

File: crates/noirc_evaluator/src/ssa/acir_gen/operations/cmp.rs (hunk 1 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/acir_gen/operations/cmp.rs:75]: No we don't, that's why I removed it. If they are equal, then x will be 0.

Comment (kevaundray) [reply crates/noirc_evaluator/src/ssa/acir_gen/operations/cmp.rs:75]: Makes sense ser, merging

**Pull request merged**